tensorboard/webapp/widgets/linked_time_fob/linked_time_fob_test.ts

```diff
@@ -322,4 +322,88 @@ describe('linked_time_fob_controller', () => {
       });
     });
   });
+
+  describe('typing step into fob', () => {
+    it('single time selection changed with fob typing', () => {
```

Comment:
You're testing the logic in linked_time_fob_controller_component, which is nice, but I think you should also test how the actual UI elements behave, too:

1) That you can double click and type into the element (testing/dom/sendKeys might be helpful here). You'd probably want to do this for both fobs.
2) That you can't type into the element when you haven't already double clicked.
3) That you can't type into the element after you've finished your input.

Reply:
I think I tested this as you said.
I am testing that you "can't type" by ensuring there is no input element.
PTAL and let me know what you think.

Reply:
> I do see that gap. Since the Fob does not know which type it is calling this would be a unit test on the controller. Somehow if we could test which type is passed down in the binding that would make sense. I do not know of a way to do that.

You already solved the hard parts. You just have to glue it together:

```
    it('changing start input modifies start step', () => {
      let fixture = createComponent({
        linkedTime: {start: {step: 1}, end: null},
      });
      fixture.detectChanges();

      let mainDiv = fixture.debugElement.query(By.css('linked-time-fob div'));
      mainDiv.triggerEventHandler('dblclick', {});
      fixture.detectChanges();

      let input = fixture.debugElement.query(By.css('linked-time-fob input'));

      sendKeys(fixture, input, '8');
      input.triggerEventHandler('change', {target: input.nativeElement});
      fixture.detectChanges();

      expect(onSelectTimeChanged).toHaveBeenCalledOnceWith({
        start: {step: 8},
        end: null,
      });
    });
```

> 
> I would also argue that testing which type is in the call when the fob calls it is not a unit test as it spans multiple components. So I think it is reasonable to only test this in the integration tests.
> 

In practice we do test this sort of thing in unit tests. We find it lighter weight to test within the Angular component framework rather than an integration test that pulls in ngrx and fake data sources. We reserve integration tests for things that are basically impossible to test only with Angular component framework. Like: 

1. Test that response received from backend modifies the UI in an expected way. (The integration here is between the data_source, effects, reducers, selectors, and Angular component layers).
2. Test that UI interaction in one part of UI modifies another part of the UI in an expected way (The integration here is between one Angular component, effects, reducers, selectors, and another Angular component).

You can consider that there are different layers of unit tests, if that helps. Some focus on a single new component (linked_time_fob_test, for example) and some that focus on a combination of new components (linked_time_fob_controller_test, for example). The latter are sort of integration tests but they are low-level, fast integration tests.

> As good testing practice suggests I also removed the LinkedTimeFobComponent from the test declarations as it should not be needed in the controller tests.

You'll have to bring it back for this particular test to work, though. I don't necessarily agree good testing practice recommends removing lower level Components or Modules. We wouldn't exclude MatButtonModule or hesitate to write a test that clicks on the button, for example. It is especially fine given that the two Components live within the same folder.

Reply:
Thanks for the full explanation! I have added 2 tests which use both components.